### ISSUE TITLE
version 2.3.0 windows, service crashing

### ISSUE THREAD

**Issue by timcanty** [NagiosEnterprises/ncpa#740]
Hello just updated a machine to the new 2.3.0 agent and after a minute or so the ncpa_listener.exe service is stopping.

This is what event viewer is showing.

![image](https://user-images.githubusercontent.com/32928117/106511954-e00b3600-64c8-11eb-9deb-7a32f3073972.png)

**Comment by jomann09** [comment]
Do you have Microsoft Visual C++ 2010 redistributable installed on the system? It should warn you when you install it, but that seems to be the component that is erroring.

**Comment by timcanty** [comment]
I would of thought so, agent 2.2.0 installed and works fine, but second put 2.3.0 on this happened.

Will try it on another machine in the morning, and see if the same issue occurs.

**Comment by jomann09** [comment]
Sounds good, I just tested it on my Windows 10 install and it isn't dying, I'll try upgrading from an old version as well. Also, what version of Windows are you running it on?

**Comment by timcanty** [comment]
One one I upgraded it on was server 2012 R2.

**Comment by timcanty** [comment]
Still no look on this server, 2.2.2. installs and works fine, 2.3.0 starts, nothing in the log other then 
"[04656] 2021/02/02 13:12:16.180 starting logging at level ERROR" and then the service crashes.
 
i've since upgraded a few other windows servers very similar spec's/ and versions and they worked, yet this one is still doing the same thing. Any further thoughts?

**Comment by ccztux** [comment]
What value has the log_level configured in ```ncpa.cfg``` section ```listener```?
```
#
# -------------------------------
# Listener Configuration (daemon)
# -------------------------------
#

[listener]

.
.
[snip]
.
.

#
# Listener logging file level, location, and the PID location
# Default: loglevel = info (debug, info, warning, error)
# Default: logfile = var/log/ncpa_listener.log
# Default: pidfile = var/run/ncpa_listener.pid (leave listener in pid file name)
#
loglevel = info
```

I would recommend to set it to debug it if isnt already.

**Comment by jomann09** [comment]
It's odd I've seen this happen before, where it doesn't work properly on a single system but works on others with the same specs. I'm going to ask some internal people to see if/how the resolved this in the past.

**Comment by jomann09** [comment]
This is documentation from Windows Server 2003 but I'm assuming it's a similar issue with DEP (https://docs.microsoft.com/en-us/previous-versions/windows/it-pro/windows-server-2003/cc738483(v=ws.10)?redirectedfrom=MSDN) that is causing the BEX error. Now I'm not sure what could fix it, but you could always try installing an older version of the visual c++ that works properly with Python 2.7 (https://www.microsoft.com/en-US/download/details.aspx?id=29)

**Comment by timcanty** [comment]
well tried the options suggested above and didn't make any difference, but after trying to install one more time, its now working. so not sure if the combination of anything above, plus a complete shut down and turned back on again.

Anyway its working, so will mark it as closed, thanks for tyring to assist.

**Comment by timcanty** [comment]
I have located another couple of vm's that are doing the same thing, the service from version 2.3.0 seems to crash when there is a mounted iso as read only floppy disk/cd drive. Its the only thing in common, Once they are removed version 2.3.0 will run fine, this didn't happen with version 2.2.2, so not sure what was changed between versions, but guessing doesn't like the read only drives.

**Comment by jomann09** [comment]
I will try to look into this further, I'm guessing it would be a psutil issue since that's what reads from that. Can you check the ncpa_listener.log file and post any errors you are getting on those machines?

**Comment by timcanty** [comment]
Nothing useful in the logs afraid, this is all it has in it (set as debug as well), 

2021-02-12 15:52:27,605:INFO:ncpa_windows:Parsed config from: ['C:\\Program Files (x86)\\Nagios\\NCPA\\etc\\ncpa.cfg', 'C:\\Program Files (x86)\\Nagios\\NCPA\\etc\\ncpa.cfg.d\\example.cfg']
2021-02-12 15:52:27,605:INFO:ncpa_windows:Looking for plugins at: C:\Program Files (x86)\Nagios\NCPA\plugins
2021-02-12 15:52:27,681:INFO:ncpa_windows:Using SSL version TLSv1_2
2021-02-12 15:52:41,611:DEBUG:handler:Initializing WebSocket
2021-02-12 15:52:41,611:DEBUG:handler:Validating WebSocket request

**Comment by SupercowsNL** [comment]
I have the same problems when using 2.3.0 and 2.3.1 on a fresh windows 2019 machine. C++ 2008 is installed.

Faulting application name: ncpa_listener.exe, version: 2.3.1.0, time stamp: 0x549debde
Faulting module name: MSVCR90.dll, version: 9.0.30729.9518, time stamp: 0x5b690a0f
Exception code: 0xc0000417
Fault offset: 0x00036bf4
Faulting process id: 0xccc
Faulting application start time: 0x01d74d57c2af26d9
Faulting application path: C:\Program Files (x86)\Nagios\NCPA\ncpa_listener.exe
Faulting module path: C:\Windows\WinSxS\x86_microsoft.vc90.crt_1fc8b3b9a1e18e3b_9.0.30729.9518_none_508db366bcbd18c4\MSVCR90.dll
Report Id: c678abe2-4be8-483b-a0a4-890e642a8f43
Faulting package full name: 
Faulting package-relative application ID: 

Version 2.2.2 is running fine

**Comment by timcanty** [comment]
Have you a got a floppy disk drive attached or read only CD-ROM drive attached by any chance to the VM, we found removing them then it doesn't crash. I didn't have time to remove them individually to work out which of the two that was causing an issue.

**Comment by SupercowsNL** [comment]
Yes I have :) It is the floppy drive! I have removed the floppy and now it is working fine. When I ejected the CD-ROM only the issue still exists

**Comment by MattTheQuick** [comment]
Not to revive an old thread, but I wanted to point out that I did a server OS wipe on a Dell PowerEdge R440 and had this same issue, which I resolved after reading the comment by @timcanty above. In my case, however, it was the removing of the mount point in Windows for the OEMDRV volume where Dell has OS drivers to assist with OS deployment. Once I did that in diskpart, my install of NCPA 2.3.1 worked as expected.

**Comment by smithgj** [comment]
Version 2.4 as well.  When a USB read-only device is plugged into a Windows system that has the NCPA listener service installed, the NCPA service will fail when a host/service check is made from the Nagios XI server. Steps To Reproduce: Install and configure NCPA listener service on a Windows machine. Plug in a read-only USB drive. Wait for (or force) a host/system check from a Nagios XI server to the Windows machine. The service can be manually restarted without rebooting the Windows machine.
This may be helpful if py2exe is used to create the plugin:
https://stackoverflow.com/questions/51844066/getting-out-of-dll-hell-with-microsoft-vc90-crt

**Comment by sdaniel-uw** [comment]
When I have a Windows installation USB drive (created using Rufus) inserted, the NCPA listener service always crashes. When I remove it, NCPA listener service immediately starts working.
